frame/contracts/src/tests.rs

```diff
@@ -2472,3 +2494,36 @@ fn deletion_queue_full() {
 		<ContractInfoOf::<Test>>::get(&addr).unwrap().get_alive().unwrap();
 	});
 }
+
+#[test]
+fn not_deployed_if_endowment_too_low_for_first_rent() {
+	let (wasm, code_hash) = compile_module::<Test>("set_rent").unwrap();
+
+	// The instantation deducted the rent for one block immediatly
+	let first_rent = <Test as Config>::RentFraction::get()
+		// base_deposit + deploy_set_storage (4 bytes in 1 item) - free_balance
+		.mul_ceil(80_000u32 + 40_000 + 10_000 - 30_000)
+		// blocks to rent
+		* 1;
+
+	ExtBuilder::default()
+		.existential_deposit(50)
+		.build()
+		.execute_with(|| {
+			// Create
+			let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+			assert_ok!(Contracts::put_code(Origin::signed(ALICE), wasm));
+			assert_err_ignore_postinfo!(Contracts::instantiate(
```

Comment:
Can you make sure this is non storage changing?

There should be a new macro `assert_storage_noop` you can wrap this in.

Reply:
By doing so I noticed that the storage root changed on failure. This was because of the trie id counter which is incremented outside of a transaction. This is OK because we only need uniqueness there and not sequence. However, I decided to wrap the increment into a new storage transaction to no modify the storage in failure.